internal/cache/safedemands.go

```diff
@@ -16,123 +16,72 @@ package cache
 
 import (
 	"context"
-	"sync"
-	"time"
 
 	demandapi "github.com/palantir/k8s-spark-scheduler-lib/pkg/apis/scaler/v1alpha1"
 	demandclient "github.com/palantir/k8s-spark-scheduler-lib/pkg/client/clientset/versioned/typed/scaler/v1alpha1"
-	ssinformers "github.com/palantir/k8s-spark-scheduler-lib/pkg/client/informers/externalversions"
 	"github.com/palantir/k8s-spark-scheduler/config"
 	"github.com/palantir/k8s-spark-scheduler/internal/crd"
-	"github.com/palantir/pkg/retry"
 	werror "github.com/palantir/witchcraft-go-error"
-	"github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log"
 	"github.com/palantir/witchcraft-go-logging/wlog/wapp"
 	"go.uber.org/atomic"
-	apiextensionsclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
-	clientcache "k8s.io/client-go/tools/cache"
-)
-
-const (
-	informerSyncRetryCount          = 5
-	informerSyncTimeout             = 2 * time.Second
-	informerSyncRetryInitialBackoff = 500 * time.Millisecond
 )
 
 // SafeDemandCache wraps a demand cache by checking if the demand
 // CRD exists before each operation
 type SafeDemandCache struct {
 	*DemandCache
 	demandCRDInitialized atomic.Bool
-	informerFactory      ssinformers.SharedInformerFactory
-	apiExtensionsClient  apiextensionsclientset.Interface
-	cacheInitialization  sync.Mutex
+	lazyDemandInformer   *crd.LazyDemandInformer
 	demandKubeClient     demandclient.ScalerV1alpha1Interface
 	asyncClientConfig    config.AsyncClientConfig
 }
 
 // NewSafeDemandCache returns a demand cache which fallbacks
 // to no-op if demand CRD doesn't exist
 func NewSafeDemandCache(
-	informerFactory ssinformers.SharedInformerFactory,
-	apiExtensionsClient apiextensionsclientset.Interface,
+	lazyDemandInformer *crd.LazyDemandInformer,
 	demandKubeClient demandclient.ScalerV1alpha1Interface,
 	asyncClientConfig config.AsyncClientConfig,
 ) *SafeDemandCache {
 	return &SafeDemandCache{
-		informerFactory:     informerFactory,
-		apiExtensionsClient: apiExtensionsClient,
-		demandKubeClient:    demandKubeClient,
-		asyncClientConfig:   asyncClientConfig,
+		lazyDemandInformer: lazyDemandInformer,
+		demandKubeClient:   demandKubeClient,
+		asyncClientConfig:  asyncClientConfig,
 	}
 }
 
 // Run starts the goroutine to check for the existence of the demand CRD
 func (sdc *SafeDemandCache) Run(ctx context.Context) {
-	if sdc.checkDemandCRDExists(ctx) {
+	err := sdc.initializeCache(ctx)
+	if err == nil {
 		return
 	}
 	go func() {
-		_ = wapp.RunWithFatalLogging(ctx, sdc.doStart)
-	}()
-}
-
-func (sdc *SafeDemandCache) doStart(ctx context.Context) error {
-	t := time.NewTicker(time.Minute)
-	defer t.Stop()
-	for {
-		select {
-		case <-ctx.Done():
-			return nil
-		case <-t.C:
-			if sdc.checkDemandCRDExists(ctx) {
-				return nil
-			}
+		err := wapp.RunWithFatalLogging(ctx, sdc.wait)
```

Comment:
I think it would be clear if this read like:

```golang
sdc.initializeCache(ctx);
go func() {
   wapp.RunWithFatalLogging(ctx, sdc.wait);
   sdc.initializeCache(ctx);
}()
```
(ignoring error handling)